lib/src/repository/default_git_branch.dart

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../logging.dart';
+import '../utils.dart' show runProc, withTempDir;
+
+final _acceptedBranchNameRegExp = RegExp(r'^[a-z0-9]+$');
+
+/// Detects the name of the default git branch on the given [baseUrl],
+/// using the metadata returned by `git remote show`.
+///
+/// Returns `null` when no branch name can be detected, or the implied
+/// branch name does not conform of the allowed patterns.
+Future<String?> tryDetectDefaultGitBranch(String baseUrl) async {
+  return await withTempDir((dir) async {
+    final initPr = await runProc(
+      ['git', 'init', dir],
+      workingDirectory: dir,
+    );
+    if (initPr.exitCode != 0) {
+      log.warning('Failed to run `git init`.\n${initPr.asJoinedOutput}');
+      return null;
+    }
+    final remoteAddPr = await runProc(
+      ['git', 'remote', 'add', 'origin', baseUrl],
+      workingDirectory: dir,
+    );
+    if (remoteAddPr.exitCode != 0) {
+      log.warning(
+          'Failed to run `git remote add`.\n${remoteAddPr.asJoinedOutput}');
+    }
+    final remoteShowPr = await runProc(
+      ['git', 'remote', 'show', 'origin'],
```

Comment:
Maybe, wrap with some retry, assuming this make network connections.

Or check in git manual that it does retry, and then leave a comment with a reference to documentation.